base/heartbeat.go

```diff
@@ -196,21 +178,69 @@ func (h *couchbaseHeartBeater) StartCheckingHeartbeats(staleThresholdMs int, han
 
 }
 
-// Stop the heartbeat checker
-func (h *couchbaseHeartBeater) StopCheckingHeartbeats() {
-	close(h.heartbeatCheckCloser)
+// Register a new HeartbeatListener.  Listeners must be registered after the heartbeater has been started,
+// to avoid the situation where a new node triggers immediate removal/rebalance because it hasn't started sending
+// heartbeats yet
+func (h *couchbaseHeartBeater) RegisterListener(handler HeartbeatListener) error {
+
+	if !h.sendActive.IsTrue() {
+		return errors.New("Heartbeater must be started before registering listeners, to avoid node removal")
+	}
+
+	h.heartbeatListenersMutex.Lock()
+	defer h.heartbeatListenersMutex.Unlock()
```

Comment:
General observation: I'm happy defer's cost now being negligible has allowed us to use them a bit more liberally. It really does tidy this type of code up!

Reply:
I agree, although I think it still makes sense to evaluate/consider usage for performance-critical paths.